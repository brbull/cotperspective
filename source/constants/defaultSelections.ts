
// Values to be used for index pages, or tests as the default page selection
export const selectedExchange = 'CHICAGO MERCANTILE EXCHANGE'
export const selectedMarket = 'EURO FX'
export const selectedTraderCategory = 'Noncommercial'
