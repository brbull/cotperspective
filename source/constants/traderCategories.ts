export const traderCategories = ['Commercial', 'Noncommercial'] as const
