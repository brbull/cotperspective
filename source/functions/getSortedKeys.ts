
/**
 * Returns a sorted array containing the keys of the given Record/Object
 */
export const getSortedKeys = (data: Record<string, unknown>): string[] => Object.keys(data).sort()
