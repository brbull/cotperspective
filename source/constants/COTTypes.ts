import { COTPropertiesToSelect } from '../constants/COTPropertiesToSelect'
import { traderCategories } from './traderCategories'

export type TraderCategory = typeof traderCategories[number]

export type CSVData = Record<typeof COTPropertiesToSelect[number], string>

export interface FormattedCSVData extends CSVData {
  market: string,
  exchange: string
}

export type MarketsData = Record<string, FormattedCSVData[]>

export type COTData = Record<string, MarketsData>
