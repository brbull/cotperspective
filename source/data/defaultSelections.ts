import { DropDownSelections } from '../components/PageDropDowns'
import { traderCategories } from './traderCategories'

/** Index pages, or test default page selection */
export const defaultSelections: DropDownSelections = {
  exchange: 'CHICAGO MERCANTILE EXCHANGE',
  market: 'EURO FX',
  traderCategory: traderCategories[1]
}
