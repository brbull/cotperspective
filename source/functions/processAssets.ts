import CleanCSS from 'clean-css'
import fs from 'fs'
import { buildPath } from '../data/buildPath'
import { execSync } from 'child_process'

const assetsPath = './source/assets'

const copyAssets = (files: string[]) => files.forEach(file =>
  fs.copyFileSync(`${assetsPath}/${file}`, `${buildPath}/${file}`))

export const processAssets = () => {
  console.log('• Copying assets')
  copyAssets(['favicon.ico', 'preview.png'])  
  console.log('• Processing styles')
  const styles = '' + fs.readFileSync('./source/assets/styles.css')
  const minStyles = new CleanCSS().minify(styles).styles
  fs.writeFileSync(`${buildPath}/styles.css`, minStyles)
}
