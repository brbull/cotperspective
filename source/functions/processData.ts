import csv from 'csv-parse/lib/sync'
import pick from 'lodash/pick'
import { propertiesToSelect } from '../data/propertiesToSelect'
import { COTData, CSVData } from '../data/types'

/**
 * Transforms csv string into js object, grouped first by exchange, then by market name.
 * This grouping makes view rendering easier (mapping through the values)
 */
export const processData = (csvContent: string): COTData =>
  csv(csvContent, { trim: true, columns: true, skip_empty_lines: true }).reduce(
    (data: COTData, csvData: CSVData) => {
      const marketNameSplit = csvData['Market and Exchange Names'].split(' - ')
      const market = marketNameSplit[0].trim()
      const exchange = marketNameSplit[1].trim()
      // Set initial values
      if (!data[exchange]) data[exchange] = {}
      if (!data[exchange][market]) data[exchange][market] = []
      // Filter csv data, and push to result
      data[exchange][market].push({
        market,
        exchange,
        ...pick(csvData, propertiesToSelect)
      })
      return data
    },
    {}
  )
