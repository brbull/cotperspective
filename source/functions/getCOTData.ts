import unzipper from 'unzipper'
import request from 'request'
import { processData } from './processData'
import { COTData } from '../constants/COTTypes'

const zipBaseURL = 'https://www.cftc.gov/sites/default/files/files/dea/history/'

/**
 * Fetch historical data from CFTC website, and parse to JSON.
 * @param year The latest year of which to load historical data
 * @param minimumEntries Minimum entries the current year should have, otherwise load more data from previous year
 */
export const getCOTData = async (year: number, minimumEntries: number): Promise<COTData> => {
  console.log(`-- fetching COT data for year: ${year}`)
  const zipURL = `${zipBaseURL}deahistfo${year}.zip`
  // Current types for unzipper seem to be incorrect
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const directory = await unzipper.Open.url(request as any, zipURL)
  const file = directory.files.find(d => d.path === 'annualof.txt')
  const contentBuffer = await file?.buffer()
  if (!contentBuffer) throw new Error('CSV content buffer undefined')
  const data = processData(contentBuffer.toString())
  // If not enough entries in current year, then load previous year, and join the data for each market
  if (data['CHICAGO MERCANTILE EXCHANGE']['EURO FX']?.length < minimumEntries) {
    console.log('-- fetching COT data for previous year')
    const previousYearData = await getCOTData(year - 1, minimumEntries)
    for (const exchange in data) {
      for (const market in data[exchange]) {
        const previousData = previousYearData[exchange][market]
        if (previousData) data[exchange][market].push(...previousData)
      }
    }
  }
  return data
}
