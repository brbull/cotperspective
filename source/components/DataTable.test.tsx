import React from 'react'
import { shallow } from 'enzyme'
import { selectedExchange, selectedMarket, selectedTraderCategory } from '../constants/defaultSelections'
import { testData } from '../constants/testData'
import { DataTable, getAverage } from './DataTable'
import { processTableData } from '../functions/processTableData'
import { averagePeriod } from '../constants/averagePeriod'

/**
 * Verify correct DataTable rendering of test data
 */
test('DataTable Snapshot', () => {
  const marketsData = testData[selectedExchange]
  const marketData = marketsData[selectedMarket]
  const dataTable = <DataTable
    averagePeriod={averagePeriod}
    values={marketData.map(processTableData(selectedTraderCategory))}
  />
  expect(shallow(dataTable)).toMatchSnapshot()
})

/**
 * Verify getAverage returns the correctly formatted average
 */
test('getAverage', () => {
  const _average = getAverage(2, [[0, 120200], [0, 320500], [0, 190510]])
  expect(_average(1)).toEqual('220,350')
})
