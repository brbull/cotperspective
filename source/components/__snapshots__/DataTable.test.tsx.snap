// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Snapshot 1`] = `
<table
  className="table table-bordered text-center"
  style={
    Object {
      "marginTop": 10,
    }
  }
>
  <thead
    className="thead-dark"
  >
    <tr>
      <th
        scope="col"
      >
        DATE
      </th>
      <th
        scope="col"
      >
        LONG
      </th>
      <th
        scope="col"
      >
        SHORT
      </th>
      <th
        scope="col"
      >
        CHANGE LONG
      </th>
      <th
        scope="col"
      >
        CHANGE SHORT
      </th>
      <th
        scope="col"
      >
        % LONG
      </th>
      <th
        scope="col"
      >
        % SHORT
      </th>
      <th
        scope="col"
      >
        NET POSITIONS
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="bg-light"
    >
      <td
        title="12 Period Simple Average"
      >
        Average(
        12
        )
      </td>
      <td>
        176,454
      </td>
      <td>
        85,405
      </td>
      <td
        colSpan={2}
      />
      <td>
        28
         %
      </td>
      <td>
        13
         %
      </td>
      <td>
        91,049
      </td>
    </tr>
    <tr
      key="0"
    >
      <td
        className=""
        key="0"
      >
        2020-07-07
      </td>
      <td
        className="column-positive"
        key="1"
      >
        186,241
      </td>
      <td
        className="column-positive"
        key="2"
      >
        80,047
      </td>
      <td
        className=""
        key="3"
      >
        5,904
      </td>
      <td
        className=""
        key="4"
      >
        1,730
      </td>
      <td
        className=""
        key="5"
      >
        28.8 %
      </td>
      <td
        className=""
        key="6"
      >
        12.4 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        106,194
      </td>
    </tr>
    <tr
      key="1"
    >
      <td
        className=""
        key="0"
      >
        2020-06-30
      </td>
      <td
        className="column-negative"
        key="1"
      >
        180,338
      </td>
      <td
        className="column-positive"
        key="2"
      >
        78,317
      </td>
      <td
        className=""
        key="3"
      >
        -10,929
      </td>
      <td
        className=""
        key="4"
      >
        6,403
      </td>
      <td
        className=""
        key="5"
      >
        28.0 %
      </td>
      <td
        className=""
        key="6"
      >
        12.2 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        102,021
      </td>
    </tr>
    <tr
      key="2"
    >
      <td
        className=""
        key="0"
      >
        2020-06-23
      </td>
      <td
        className="column-positive"
        key="1"
      >
        191,267
      </td>
      <td
        className="column-positive"
        key="2"
      >
        71,914
      </td>
      <td
        className=""
        key="3"
      >
        3,162
      </td>
      <td
        className=""
        key="4"
      >
        2,053
      </td>
      <td
        className=""
        key="5"
      >
        29.5 %
      </td>
      <td
        className=""
        key="6"
      >
        11.1 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        119,353
      </td>
    </tr>
    <tr
      key="3"
    >
      <td
        className=""
        key="0"
      >
        2020-06-16
      </td>
      <td
        className="column-positive"
        key="1"
      >
        188,105
      </td>
      <td
        className="column-negative"
        key="2"
      >
        69,861
      </td>
      <td
        className=""
        key="3"
      >
        1,305
      </td>
      <td
        className=""
        key="4"
      >
        -19,391
      </td>
      <td
        className=""
        key="5"
      >
        29.5 %
      </td>
      <td
        className=""
        key="6"
      >
        11.0 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        118,244
      </td>
    </tr>
    <tr
      key="4"
    >
      <td
        className=""
        key="0"
      >
        2020-06-09
      </td>
      <td
        className="column-positive"
        key="1"
      >
        186,800
      </td>
      <td
        className="column-negative"
        key="2"
      >
        89,251
      </td>
      <td
        className=""
        key="3"
      >
        12,662
      </td>
      <td
        className=""
        key="4"
      >
        -2,579
      </td>
      <td
        className=""
        key="5"
      >
        28.5 %
      </td>
      <td
        className=""
        key="6"
      >
        13.6 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        97,549
      </td>
    </tr>
    <tr
      key="5"
    >
      <td
        className=""
        key="0"
      >
        2020-06-02
      </td>
      <td
        className="column-negative"
        key="1"
      >
        174,138
      </td>
      <td
        className="column-negative"
        key="2"
      >
        91,831
      </td>
      <td
        className=""
        key="3"
      >
        -105
      </td>
      <td
        className=""
        key="4"
      >
        -5,732
      </td>
      <td
        className=""
        key="5"
      >
        26.9 %
      </td>
      <td
        className=""
        key="6"
      >
        14.2 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        82,307
      </td>
    </tr>
    <tr
      key="6"
    >
      <td
        className=""
        key="0"
      >
        2020-05-26
      </td>
      <td
        className="column-positive"
        key="1"
      >
        174,243
      </td>
      <td
        className="column-positive"
        key="2"
      >
        97,562
      </td>
      <td
        className=""
        key="3"
      >
        7,524
      </td>
      <td
        className=""
        key="4"
      >
        3,817
      </td>
      <td
        className=""
        key="5"
      >
        28.6 %
      </td>
      <td
        className=""
        key="6"
      >
        16.0 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        76,681
      </td>
    </tr>
    <tr
      key="7"
    >
      <td
        className=""
        key="0"
      >
        2020-05-19
      </td>
      <td
        className="column-negative"
        key="1"
      >
        166,719
      </td>
      <td
        className="column-positive"
        key="2"
      >
        93,745
      </td>
      <td
        className=""
        key="3"
      >
        -3,425
      </td>
      <td
        className=""
        key="4"
      >
        1,257
      </td>
      <td
        className=""
        key="5"
      >
        27.7 %
      </td>
      <td
        className=""
        key="6"
      >
        15.6 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        72,974
      </td>
    </tr>
    <tr
      key="8"
    >
      <td
        className=""
        key="0"
      >
        2020-05-12
      </td>
      <td
        className="column-positive"
        key="1"
      >
        170,143
      </td>
      <td
        className="column-positive"
        key="2"
      >
        92,488
      </td>
      <td
        className=""
        key="3"
      >
        4,569
      </td>
      <td
        className=""
        key="4"
      >
        886
      </td>
      <td
        className=""
        key="5"
      >
        28.4 %
      </td>
      <td
        className=""
        key="6"
      >
        15.5 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        77,655
      </td>
    </tr>
    <tr
      key="9"
    >
      <td
        className=""
        key="0"
      >
        2020-05-05
      </td>
      <td
        className="column-positive"
        key="1"
      >
        165,575
      </td>
      <td
        className="column-positive"
        key="2"
      >
        91,602
      </td>
      <td
        className=""
        key="3"
      >
        763
      </td>
      <td
        className=""
        key="4"
      >
        5,891
      </td>
      <td
        className=""
        key="5"
      >
        27.4 %
      </td>
      <td
        className=""
        key="6"
      >
        15.1 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        73,973
      </td>
    </tr>
    <tr
      key="10"
    >
      <td
        className=""
        key="0"
      >
        2020-04-28
      </td>
      <td
        className="column-negative"
        key="1"
      >
        164,811
      </td>
      <td
        className="column-positive"
        key="2"
      >
        85,712
      </td>
      <td
        className=""
        key="3"
      >
        -4,256
      </td>
      <td
        className=""
        key="4"
      >
        3,182
      </td>
      <td
        className=""
        key="5"
      >
        27.1 %
      </td>
      <td
        className=""
        key="6"
      >
        14.1 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        79,099
      </td>
    </tr>
    <tr
      key="11"
    >
      <td
        className=""
        key="0"
      >
        2020-04-21
      </td>
      <td
        className="column-positive"
        key="1"
      >
        169,068
      </td>
      <td
        className="column-positive"
        key="2"
      >
        82,530
      </td>
      <td
        className=""
        key="3"
      >
        5,425
      </td>
      <td
        className=""
        key="4"
      >
        4,086
      </td>
      <td
        className=""
        key="5"
      >
        27.8 %
      </td>
      <td
        className=""
        key="6"
      >
        13.6 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        86,538
      </td>
    </tr>
    <tr
      key="12"
    >
      <td
        className=""
        key="0"
      >
        2020-04-14
      </td>
      <td
        className="column-positive"
        key="1"
      >
        163,643
      </td>
      <td
        className="column-negative"
        key="2"
      >
        78,444
      </td>
      <td
        className=""
        key="3"
      >
        2,440
      </td>
      <td
        className=""
        key="4"
      >
        -4,657
      </td>
      <td
        className=""
        key="5"
      >
        27.2 %
      </td>
      <td
        className=""
        key="6"
      >
        13.1 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        85,199
      </td>
    </tr>
    <tr
      key="13"
    >
      <td
        className=""
        key="0"
      >
        2020-04-07
      </td>
      <td
        className="column-positive"
        key="1"
      >
        161,204
      </td>
      <td
        className="column-positive"
        key="2"
      >
        83,102
      </td>
      <td
        className=""
        key="3"
      >
        6,022
      </td>
      <td
        className=""
        key="4"
      >
        832
      </td>
      <td
        className=""
        key="5"
      >
        26.2 %
      </td>
      <td
        className=""
        key="6"
      >
        13.5 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        78,102
      </td>
    </tr>
    <tr
      key="14"
    >
      <td
        className=""
        key="0"
      >
        2020-03-31
      </td>
      <td
        className="column-positive"
        key="1"
      >
        155,182
      </td>
      <td
        className="column-negative"
        key="2"
      >
        82,269
      </td>
      <td
        className=""
        key="3"
      >
        5,445
      </td>
      <td
        className=""
        key="4"
      >
        -11,733
      </td>
      <td
        className=""
        key="5"
      >
        24.4 %
      </td>
      <td
        className=""
        key="6"
      >
        13.0 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        72,913
      </td>
    </tr>
    <tr
      key="15"
    >
      <td
        className=""
        key="0"
      >
        2020-03-24
      </td>
      <td
        className="column-positive"
        key="1"
      >
        149,736
      </td>
      <td
        className="column-negative"
        key="2"
      >
        94,002
      </td>
      <td
        className=""
        key="3"
      >
        1,583
      </td>
      <td
        className=""
        key="4"
      >
        -23,066
      </td>
      <td
        className=""
        key="5"
      >
        22.5 %
      </td>
      <td
        className=""
        key="6"
      >
        14.1 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        55,734
      </td>
    </tr>
    <tr
      key="16"
    >
      <td
        className=""
        key="0"
      >
        2020-03-17
      </td>
      <td
        className="column-negative"
        key="1"
      >
        148,153
      </td>
      <td
        className="column-negative"
        key="2"
      >
        117,068
      </td>
      <td
        className=""
        key="3"
      >
        -4,126
      </td>
      <td
        className=""
        key="4"
      >
        -47,647
      </td>
      <td
        className=""
        key="5"
      >
        21.8 %
      </td>
      <td
        className=""
        key="6"
      >
        17.2 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        31,085
      </td>
    </tr>
    <tr
      key="17"
    >
      <td
        className=""
        key="0"
      >
        2020-03-10
      </td>
      <td
        className="column-negative"
        key="1"
      >
        152,279
      </td>
      <td
        className="column-negative"
        key="2"
      >
        164,716
      </td>
      <td
        className=""
        key="3"
      >
        -868
      </td>
      <td
        className=""
        key="4"
      >
        -72,432
      </td>
      <td
        className=""
        key="5"
      >
        20.2 %
      </td>
      <td
        className=""
        key="6"
      >
        21.8 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        -12,437
      </td>
    </tr>
    <tr
      key="18"
    >
      <td
        className=""
        key="0"
      >
        2020-03-03
      </td>
      <td
        className="column-negative"
        key="1"
      >
        153,147
      </td>
      <td
        className="column-negative"
        key="2"
      >
        237,148
      </td>
      <td
        className=""
        key="3"
      >
        -324
      </td>
      <td
        className=""
        key="4"
      >
        -40,089
      </td>
      <td
        className=""
        key="5"
      >
        20.4 %
      </td>
      <td
        className=""
        key="6"
      >
        31.5 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        -84,001
      </td>
    </tr>
    <tr
      key="19"
    >
      <td
        className=""
        key="0"
      >
        2020-02-25
      </td>
      <td
        className="column-negative"
        key="1"
      >
        153,472
      </td>
      <td
        className="column-positive"
        key="2"
      >
        277,237
      </td>
      <td
        className=""
        key="3"
      >
        -15,222
      </td>
      <td
        className=""
        key="4"
      >
        6,681
      </td>
      <td
        className=""
        key="5"
      >
        21.6 %
      </td>
      <td
        className=""
        key="6"
      >
        39.0 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        -123,765
      </td>
    </tr>
    <tr
      key="20"
    >
      <td
        className=""
        key="0"
      >
        2020-02-18
      </td>
      <td
        className="column-positive"
        key="1"
      >
        168,693
      </td>
      <td
        className="column-positive"
        key="2"
      >
        270,555
      </td>
      <td
        className=""
        key="3"
      >
        313
      </td>
      <td
        className=""
        key="4"
      >
        10,686
      </td>
      <td
        className=""
        key="5"
      >
        24.2 %
      </td>
      <td
        className=""
        key="6"
      >
        38.8 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        -101,862
      </td>
    </tr>
    <tr
      key="21"
    >
      <td
        className=""
        key="0"
      >
        2020-02-11
      </td>
      <td
        className="column-positive"
        key="1"
      >
        168,381
      </td>
      <td
        className="column-positive"
        key="2"
      >
        259,869
      </td>
      <td
        className=""
        key="3"
      >
        682
      </td>
      <td
        className=""
        key="4"
      >
        12,193
      </td>
      <td
        className=""
        key="5"
      >
        25.0 %
      </td>
      <td
        className=""
        key="6"
      >
        38.6 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        -91,488
      </td>
    </tr>
    <tr
      key="22"
    >
      <td
        className=""
        key="0"
      >
        2020-02-04
      </td>
      <td
        className="column-negative"
        key="1"
      >
        167,699
      </td>
      <td
        className="column-positive"
        key="2"
      >
        247,676
      </td>
      <td
        className=""
        key="3"
      >
        -10,186
      </td>
      <td
        className=""
        key="4"
      >
        3,693
      </td>
      <td
        className=""
        key="5"
      >
        25.7 %
      </td>
      <td
        className=""
        key="6"
      >
        38.0 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        -79,977
      </td>
    </tr>
    <tr
      key="23"
    >
      <td
        className=""
        key="0"
      >
        2020-01-28
      </td>
      <td
        className="column-positive"
        key="1"
      >
        177,885
      </td>
      <td
        className="column-positive"
        key="2"
      >
        243,983
      </td>
      <td
        className=""
        key="3"
      >
        12,606
      </td>
      <td
        className=""
        key="4"
      >
        26,679
      </td>
      <td
        className=""
        key="5"
      >
        27.2 %
      </td>
      <td
        className=""
        key="6"
      >
        37.2 %
      </td>
      <td
        className="column-negative"
        key="7"
      >
        -66,098
      </td>
    </tr>
    <tr
      key="24"
    >
      <td
        className=""
        key="0"
      >
        2020-01-21
      </td>
      <td
        className="column-positive"
        key="1"
      >
        165,279
      </td>
      <td
        className="column-positive"
        key="2"
      >
        217,304
      </td>
      <td
        className=""
        key="3"
      >
        3,396
      </td>
      <td
        className=""
        key="4"
      >
        2,873
      </td>
      <td
        className=""
        key="5"
      >
        26.3 %
      </td>
      <td
        className=""
        key="6"
      >
        34.6 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        -52,025
      </td>
    </tr>
    <tr
      key="25"
    >
      <td
        className=""
        key="0"
      >
        2020-01-14
      </td>
      <td
        className="column-negative"
        key="1"
      >
        161,883
      </td>
      <td
        className="column-negative"
        key="2"
      >
        214,431
      </td>
      <td
        className=""
        key="3"
      >
        -2,134
      </td>
      <td
        className=""
        key="4"
      >
        -14,275
      </td>
      <td
        className=""
        key="5"
      >
        26.3 %
      </td>
      <td
        className=""
        key="6"
      >
        34.9 %
      </td>
      <td
        className="column-positive"
        key="7"
      >
        -52,548
      </td>
    </tr>
    <tr
      key="26"
    >
      <td
        className=""
        key="0"
      >
        2020-01-07
      </td>
      <td
        className=""
        key="1"
      >
        164,017
      </td>
      <td
        className=""
        key="2"
      >
        228,706
      </td>
      <td
        className=""
        key="3"
      >
        -1,827
      </td>
      <td
        className=""
        key="4"
      >
        -12,365
      </td>
      <td
        className=""
        key="5"
      >
        27.5 %
      </td>
      <td
        className=""
        key="6"
      >
        38.3 %
      </td>
      <td
        className=""
        key="7"
      >
        -64,689
      </td>
    </tr>
  </tbody>
</table>
`;
